# CodFISH
**Cod**on **F**requency **I**ndexing **S**oftware for **H**IV

The HIVDB Sequence Reads Interpretation Program accepts a codon frequency table that stores in the **CodFISH** format. The CodFISH format consists of five columns:

1. gene (`PR`, `RT`, or `IN`);
2. position;
3. total number of reads of this position;
4. codon nucleotide triplet; and
5. total number of reads of this codon.

## Usage

### Create `.codfish` file from `.fastq` file

1. Install Docker CE (https://docs.docker.com/install/).

2. Download script:

   ```bash
   sudo curl -sL https://raw.githubusercontent.com/hivdb/codfish/master/bin/fastq2codfish-docker -o /usr/local/bin/fastq2codfish
   chmod +x /usr/local/bin/fastq2codfish
   ```

3. Use following command to process FASTQ files and generate CodFISH files.

   ```bash
   fastq2codfish /path/to/folders/containing/fastq/files
   ```

   The script will automatically find every file named with an extension of `.fastq`, align them to `.sam` file and then extract
   the codon freqency table into `.codfish` file.
   
   The above command is adequate for most case of both paired or unpaired FASTQ files generated by Illumina with the filename
   pattern look like `*_L001_R1_001.fastq.gz` and `*_L001_R1_002.fastq.gz`. However, if your FSTAQ files are in other naming
   convention, please read [advanced usages#change naming convention](#change.naming.convention).

Note: the `fastq2codfish` script can only be executed in an Unix-like system. If you are using Microsoft Windows 10,
you need to install the [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install-win10) to
use this script.

#### Advanced usages
##### Change naming convention
You can specify your FASTQ file naming convention by passing `-p <PATTERN>`, `-r <REPLACE>`, `-1 <PAIR1_SUFFIX>` and 
`-2 <PAIR2_SUFFIX>` parameters to `fastq2codfish`. Noted `-p` and `-r` are paired regular expression replacement.
For example, if your FASTQ files are downloaded from Sequence Reads Archive (SRA), their naming convention would be
`*_1.fastq.gz` and `*_2.fastq.gz`. Here is the command:

```bash
fastq2codfish -p '(.+)_[12]\.fastq\.gz$' -r '\1' -1 '_1.fastq.gz' -2 '_2.fastq.gz' /path/to/folders/containing/fastq/files
```

#### Specify concurrency rate
You can also specify concurrency rate (how many CPUs you want to use) by passing `-n <NTHREADS>` to `fastq2codfish`.
By default, the script will use fomula `FLOOR(TOTAL_CPUS * 19 / 20)` to calculate the `NTHREADS` value.

```bash
# Use at most 5 CPUs
fastq2codfish -n 5 /path/to/folders/containing/fastq/files
```

Note: if you are using Docker on Windows or MacOS, the upper limit of `NTHREADS` depends on the number of CPUs allocated
for the Docker software. Please check these general answers from Stackoverflow:

- Windows: https://stackoverflow.com/a/56583203/2644759
- MacOS: https://stackoverflow.com/a/39720010/2644759

### Convert BAM/SAM file to codfish/nucfish format file

1. Install Docker CE (https://docs.docker.com/install/).

2. Download script:
   For exporting "nucfish" format, download "sam2nucfish-docker":

   ```bash
   curl -sL https://raw.githubusercontent.com/hivdb/codfish/master/bin/sam2nucfish-docker -o sam2nucfish-docker
   chmod +x sam2nucfish-docker
   ```

   For exporting "codfish" format, download "sam2codfish-docker":

   ```bash
   curl -sL https://raw.githubusercontent.com/hivdb/codfish/master/bin/sam2codfish-docker -o sam2codfish-docker
   chmod +x sam2codfish-docker
   ```

3. Use following command to convert BAM/SAM files.

   nucfish:
   ```bash
   ./sam2nucfish-docker /path/to/folders/containing/bam/files
   ```

   codfish:
   ```bash
   ./sam2codfish-docker /path/to/folders/containing/bam/files
   ```

   The script will automatically find all files with extension of .bam or .sam and convert them.

#### Troubleshoots

1. Error "`Padding (BAM_CPAD, X) is currently not supported. Please implement. Sorry about that.`"

   If you are using Geneious to export the BAM file, uncheck the "export padded CIGARs" option.

2. Error "`fetch called on bamfile without index`"

   If you are using Geneious to export the BAM file, check the "export BAM index file" option with the file extension ".bai".
